bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesReferenceProvider.java

```diff
@@ -0,0 +1,192 @@
+        }
+        return componentResourceTypes;
+    }
+
+
+    /**
+     * Iterates through the given collection of resourceTypes and finds all the components that has
+     * shared dialog, then adds the root page of the given resource to the list.
+     *
+     * @param resource {@link Resource}
+     * @param componentResourceTypes {@link Set<String>}
+     * @return references
+     */
+    private List<Reference> findSharedComponents(Resource resource,
+        Set<String> componentResourceTypes) {
+        List<Reference> references = new ArrayList<>();
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
+        for (String resourceType : componentResourceTypes) {
+
+            Resource componentResource = getComponentResource(resourceType, resourceResolver);
+            if (componentResource == null) {
+                continue;
+            }
+
+            String jcrPrimaryType = componentResource.getValueMap().get(JcrConstants.JCR_PRIMARYTYPE, String.class);
+            if (StringUtils.isEmpty(jcrPrimaryType) || !jcrPrimaryType.equalsIgnoreCase("cq:Component")) {
+                continue;
+            }
+
+            Resource sharedDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogshared");
+            Resource globalDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogglobal");
+            if (sharedDialog == null && globalDialog == null) {
+                continue;
+            }
+
+            Page rootPage = pageRootProvider.getRootPage(resource);
+            if (rootPage != null) {
+                references.add(new Reference("sharedProperties", getReferenceName(rootPage),
+                    rootPage.getContentResource(),
+                    rootPage.getLastModified() != null ? rootPage.getLastModified()
+                        .getTimeInMillis() : -1L));
+
+                return references;
+            }
+        }
+        return references;
+    }
+
+    /**
+     * Returns the Component Resource by searching the configured resolver search paths ["/libs", "/apps"]
+     * @param resourceType {@link String}
+     * @param resourceResolver {@link ResourceResolver}
+     * @return Resource
+     */
+    private Resource getComponentResource(String resourceType, ResourceResolver resourceResolver) {
+        String[] resolverSearchPaths = resourceResolver.getSearchPath();
+
+        for (String searchPath : resolverSearchPaths) {
+            Resource resource = resourceResolver.getResource(searchPath + resourceType);
+            if (resource != null) {
+                return resource;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get the page title prop if it's null get the name
+     *
+     * @param page {@link Page}
+     * @return page title | name
+     */
+    private String getReferenceName(@Nonnull Page page) {
```

Comment:
Technically speaking, could probably eliminate this function since the entire logic is a one-liner and is never repeated.  However, I'm cool if you think this is best.